---
title: "AllelicImbalance Vignette"
author: "Jesper R. GÃ¥din and Lasse Folkersen"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

This AllelicImbalance package contains functions for investigating
allelic imbalance effects in RNA-seq data. Maternal and paternal
alleles could be expected to show identical transcription rate, resulting in a
50%-50% mix of maternal and paternal mRNA in a sample. However, this turns out
to sometimes not be the case. The most extreme example is the X-chromosome
inactivation in females, but many autosomal genes also have deviations from
identical transcription rate. The causes of this are not always known, but one
likely cause is the difference in DNA, namely heterozygous SNPs, affecting
enhancers, promoter regions, splicing and stability. Identifying this allelic
imbalance is therefore of interest to the characterization of the genome and the
aim of the AllelicImbalance package is to facilitate this.

## Simple example of building an ASEset object
In this section we will walk through the various ways an ASEset
object can be created. The ASEset object has the SummarizedExperiment as parent class, and all functions you can apply on this class you can also apply on an ASEset. Although the preprocessing of RNA-seq data is not the
primary focus of this package, it is a necessary step before analysis. There
exists several different methods for obtaining a bam file, and this section
should just be considered an example. For further details we refer
to the web-pages of tophat, bowtie, bwa and samtools found in the links section at the end of this document.

```
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR009/ERR009135/*
bowtie -q --best --threads 5 --sam hg19 +
>      -1 ERR009135_1.fastq.gz -2 ERR009135_2.fastq.gz "ERR009135.sam"
samtools view -S -b ERR009135.sam > ERR009135.bam
```

In the above code one paired-end RNA sequencing sample is downloaded and
aligned to the human genome, then converted to bam using samtools. The
resulting bam files can be the direct input to the AllelicImbalance package. Other
aligners can be used as well, as long as bam files are provided as input. The
example code following illustrates how to use the import mechanism on a
chromosome 17-located subset of 20 RNA-seq experiments of HapMap samples. The
output is an \Robject{ASEset} object containing allele counts for all
heterozygote coding SNPs in the region.

> "The fact that nature provides the techniques to study itself"
